### PR TITLE
[llvm][CodeGen] Fix failure in window scheduler caused by phi

### PR DESCRIPTION
**Pull request by huaatian**
In certain cases, the register passed with the kernel MBB in phi are not defined 
within the kernel MBB. This patch adds the corresponding handling.

### CODE REVIEW

**Comment by huaatian** [comment]
The cause of the issue is that in the phi instruction, the register passed along with the kernel MBB is not defined within the kernel MBB. This scenario is uncommon, and the window algorithm currently does not support this structure. Therefore, a check is performed at the beginning of the algorithm.
image

**Comment by huaatian** [comment]
Could you please review this code? Thank you~ @nathanchance @arsenm @dtcxzyw

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 1]
```diff
@@ -198,6 +198,12 @@ bool WindowScheduler::initialize() {
     if (MI.isMetaInstruction() || MI.isTerminator())
       continue;
     if (MI.isPHI()) {
+      if (Register AntiReg = getAntiRegister(&MI))
+        // Register with Kernel in phi is not defined within the Kernel itself.
+        if (MRI->getVRegDef(AntiReg)->getParent() != MBB) {
+          LLVM_DEBUG(dbgs() << "Special phi structure is not supported!\n");
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:204]: It's not really special. Why is this difficult to handle? It should behave like any other live in value

File: llvm/test/CodeGen/Hexagon/swp-ws-fail-3.mir [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    S2_storerb_io %0, 0, killed %6
+    S4_storerb_rr %0, %4, 0, %2
+    %5:intregs = A2_tfrsi 1
+    ENDLOOP0 %bb.2, implicit-def $pc, implicit-def $lc0, implicit $sa0, implicit $lc0
+    J2_jump %bb.1, implicit-def $pc
+
+...
+
+
+
```
Comment (arsenm) [inline llvm/test/CodeGen/Hexagon/swp-ws-fail-3.mir:44]: Extra blank lines

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 3]
```diff
@@ -198,6 +198,12 @@ bool WindowScheduler::initialize() {
     if (MI.isMetaInstruction() || MI.isTerminator())
       continue;
     if (MI.isPHI()) {
+      if (Register AntiReg = getAntiRegister(&MI))
+        // Register with Kernel in phi is not defined within the Kernel itself.
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:202]: By kernel do you mean MachineBasicBlock?

File: llvm/test/CodeGen/Hexagon/swp-ws-fail-3.mir (hunk 2 above)
Comment (huaatian) [reply llvm/test/CodeGen/Hexagon/swp-ws-fail-3.mir]: Updated

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 3 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp]: Updated

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 1 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp]: Yes, you are right. We only need to add a small amount of checks.
Updated.

**Review by arsenm — APPROVED** [review]

**Pull request merged**